## 1 Find id, name and term of the artist who played the track with id TRMMWLD128F9301BF2

In [4]:
import sys
import sqlite3
with sqlite3.connect('Sample_Song_Dataset.db') as conn:
    cursor = conn.cursor()
    cursor.execute('''SELECT artists.artist_id, artists.artist_name, artist_term.term 
                      From artists 
                      JOIN artist_term 
                      ON artists.artist_id = artist_term.artist_id 
                      JOIN tracks 
                      ON tracks.artist_id = artists.artist_id 
                      WHERE tracks.track_id ="TRMMWLD128F9301BF2"''')
    print(cursor.fetchall())

[('ARE592P1187B99F3A7', 'Tone X', 'alternative rock'), ('ARE592P1187B99F3A7', 'Tone X', 'american'), ('ARE592P1187B99F3A7', 'Tone X', 'black gospel'), ('ARE592P1187B99F3A7', 'Tone X', 'black music'), ('ARE592P1187B99F3A7', 'Tone X', 'christian'), ('ARE592P1187B99F3A7', 'Tone X', 'contemporary christian'), ('ARE592P1187B99F3A7', 'Tone X', 'contemporary gospel'), ('ARE592P1187B99F3A7', 'Tone X', 'dance'), ('ARE592P1187B99F3A7', 'Tone X', 'electronic'), ('ARE592P1187B99F3A7', 'Tone X', 'funk'), ('ARE592P1187B99F3A7', 'Tone X', 'funk soul'), ('ARE592P1187B99F3A7', 'Tone X', 'gospel'), ('ARE592P1187B99F3A7', 'Tone X', 'gospel jams'), ('ARE592P1187B99F3A7', 'Tone X', 'gospel soul'), ('ARE592P1187B99F3A7', 'Tone X', 'hip hop'), ('ARE592P1187B99F3A7', 'Tone X', 'house'), ('ARE592P1187B99F3A7', 'Tone X', 'jazz'), ('ARE592P1187B99F3A7', 'Tone X', 'latin'), ('ARE592P1187B99F3A7', 'Tone X', 'male vocalist'), ('ARE592P1187B99F3A7', 'Tone X', 'pop'), ('ARE592P1187B99F3A7', 'Tone X', 'r&b'), ('ARE592

## 2. Select all the unique tracks with the duration is strictly greater than 3020 seconds.

In [9]:
cursor.execute('''SELECT DISTINCT * 
                  FROM tracks 
                  WHERE duration>"3020"''')
print(cursor.fetchall())

[('TRBOGZY128F4269C26', 'Hauptteil', 'Endlich Schlank', 0, 3020.87791, 'ARRXCAY1242077F342'), ('TRDZTTO12903CF1A2E', 'Grounation', 'Carried Beyond', 0, 3034.90567, 'ARSSDRQ1241B9CB1A5'), ('TRGASQU128F92D762F', 'Rag: Gaud Malhar & Rag: Mian Malhar', 'Nidhi', 0, 3030.17751, 'ARG62WR1187FB461BC'), ('TRGXALK12903D15504', 'Don Carlos: Act III', 'Verdi: Don Carlos', 0, 3024.97914, 'ARFULUG1272BA849E4'), ('TRHWGWD12903CD132B', 'Germans are people too (Alive & Kickin)', 'Germans Are People too (Alive & Kickin´)', 0, 3027.17342, 'ARFIKPQ1272BCD1FD4'), ('TRJOADI12903CC00F9', 'Monolithe II', 'Monolithe II', 2005, 3024.66567, 'AR8XMK11187B9AF8C7'), ('TRLOKWE12903CF205E', 'Hud', 'Le Saint Coran (avec Doae)', 0, 3026.75546, 'ARHDNFW11F4C83BECD'), ('TRMTJMT128F93221B0', 'Galactic Odyssey', 'Galactic Odyssey', 0, 3029.08036, 'ARNHB3M1187B98B512'), ('TROVFHS128F934D39D', 'Live Acid Wandererz', 'Live Acid Wandererz', 0, 3026.57261, 'ARIAXFE11F50C50923'), ('TROZPGO128F426FD26', 'Kunverbainu Mameru - Aakh

## 3. Find the ten shortest (by duration) 10 tracks released between 2010 and 2014 (inclusive), ordered by increasing duration.

In [10]:
cursor.execute('''SELECT * 
                  FROM tracks 
                  WHERE year 
                  BETWEEN "2010" AND "2014" 
                  ORDER BY duration 
                  LIMIT 10''')
print(cursor.fetchall())

[('TRUJRNE12903CCE156', 'Spirit', 'Deep & Meaningless Remixes', 2010, 3.70893, 'ARIWIEL1187B9A2197'), ('TRLMWPC12903D0EB6D', '64:14', 'Audio_Data_Bank', 2010, 5.0673, 'ARYLUL51187B9947D9'), ('TRPMUAO12903CBBEA3', 'Fire With Fire', 'Hanging On By A Thread', 2010, 6.66077, 'ARZKRMD1241B9C9F21'), ('TRTKXQG12903D06EA3', 'Certain Things (Interlude)', 'I\x19m New Here', 2010, 8.48934, 'AREPYZO1187B9B9C2D'), ('TRCERDU12903D03CF7', 'Night of the Morbid Psycho', 'Psalm of the Grand Destroyer', 2010, 10.86649, 'ARB42181187B9A4656'), ('TRGVKMD12903CEB193', 'Auto', 'Bilder Fressen Strom', 2010, 11.4673, 'ARLB90U1187B996AE5'), ('TRXWWHJ12903CAB222', "Pablo's Heart", 'There Is Love In You', 2010, 11.85914, 'AROKAO61187FB38105'), ('TRVJQOT12903CEB184', 'Goldene Mitte', 'Bilder Fressen Strom', 2010, 12.56444, 'ARLB90U1187B996AE5'), ('TRADDAQ12903CBBEA4', 'Invisible', 'Hanging On By A Thread', 2010, 14.13179, 'ARZKRMD1241B9C9F21'), ('TRJUJCN12903D06E9F', 'I Was Guided (Interlude)', 'I\x19m New Here', 2

## 4. Find the top 20 most frequently used terms, ordered by decreasing usage.

In [11]:
cursor.execute('''SELECT term 
                  FROM artist_term 
                  GROUP BY term 
                  ORDER BY COUNT(*) DESC 
                  LIMIT 20''')
print(cursor.fetchall())

[('rock',), ('electronic',), ('pop',), ('alternative rock',), ('hip hop',), ('united states',), ('house',), ('jazz',), ('indie',), ('alternative',), ('electro',), ('experimental',), ('indie rock',), ('pop rock',), ('punk',), ('techno',), ('folk',), ('downtempo',), ('ambient',), ('soul',)]


## 5. Find the artist name associated with the longest track duration.

In [12]:
#cursor.execute('SELECT artists.artist_name FROM artists JOIN tracks on tracks.artist_id = artists.artist_id ORDER BY tracks.duration DESC LIMIT 1')
cursor.execute('''SELECT artists.artist_name 
                  FROM artists 
                  JOIN tracks 
                  ON tracks.artist_id = artists.artist_id 
                  WHERE duration=(SELECT MAX(duration) FROM tracks)''')
print(cursor.fetchall())

[('Mystic Revelation of Rastafari',)]


## 6. Find the mean duration of all tracks.

In [13]:
cursor.execute('''SELECT AVG(duration) 
                  FROM tracks''')
print(cursor.fetchall())

[(249.71338394213936,)]


## 7. Using only one query, count the number of tracks whose artists don't have any linked terms.

In [14]:
cursor.execute('''SELECT COUNT(*) 
                  FROM tracks 
                  LEFT JOIN artist_term 
                  ON tracks.artist_id = artist_term.artist_id 
                  WHERE artist_term.artist_id IS NULL''')
#cursor.execute('SELECT COUNT(*) FROM tracks WHERE tracks.artist_id NOT IN (SELECT artist_id FROM artist_term)')
print(cursor.fetchall())



[(3710,)]


## 8. Index- Run Question 1 query in a loop for 100 times and note the minimum time taken. Now create an index on the column artist_id and compare the time. Share your findings in the report.

In [15]:
import time
start = time.time()
for i in range(100):
    cursor.execute('''SELECT artists.artist_id, artists.artist_name, artist_term.term 
                      From artists 
                      JOIN artist_term 
                      ON artists.artist_id = artist_term.artist_id 
                      JOIN tracks 
                      ON tracks.artist_id = artists.artist_id 
                      WHERE tracks.track_id ="TRMMWLD128F9301BF2"''')
end = time.time()
print(end-start)

cursor.execute('''CREATE INDEX artist_id_index_tracks 
                ON tracks (artist_id)''')
cursor.execute('''CREATE INDEX artist_id_index_artists 
                  ON artists (artist_id)''')
cursor.execute('''CREATE INDEX artist_id_index_artist_term 
                  ON artist_term (artist_id)''')
start1 = time.time()
for i in range(100):
    cursor.execute('''SELECT artists.artist_id, artists.artist_name, artist_term.term 
                      From artists 
                      JOIN artist_term 
                      ON artists.artist_id = artist_term.artist_id 
                      JOIN tracks 
                      ON tracks.artist_id = artists.artist_id 
                      WHERE tracks.track_id ="TRMMWLD128F9301BF2"''')
end1 = time.time()
print(end1-start1)

148.9804871082306
0.0026738643646240234


Before creating the index, the program runs for 149 seconds. After adding index, the running time is only 0.0027 seconds, which are much faster. Index helps us joining table more efficiently in terms of searching for the same value for the primary key

## 9. Find all tracks associated with artists that have the tag eurovision winner and delete them from the database, then roll back this query using a transaction. Hint: you can select from the output of a select!

In [16]:
cursor.execute('BEGIN TRANSACTION')
cursor.execute('''DELETE FROM tracks 
                  WHERE tracks.track_id 
                  IN (SELECT tracks.track_id 
                  FROM tracks JOIN artist_term on artist_term.artist_id = tracks.artist_id 
                  WHERE artist_term.term ="eurovision winner")''')
cursor.execute('ROLLBACK')
print(cursor.fetchall())

# Check if it's rolled back
cursor.execute('SELECT * FROM tracks JOIN artist_term on artist_term.artist_id = tracks.artist_id WHERE artist_term.term ="eurovision winner"')
print(cursor.fetchall())

[]
[('TRMBQRR128F93036AA', 'Die weißen Segel', 'Sag mir wo wohnen die Engel', 0, 193.09669, 'ARAGY2M1187B9A6712', 'ARAGY2M1187B9A6712', 'eurovision winner'), ('TRGBZAK128F4252004', 'So wie du denkst so wird es kommen', 'Sag mir wo wohnen die Engel', 0, 225.72363, 'ARAGY2M1187B9A6712', 'ARAGY2M1187B9A6712', 'eurovision winner'), ('TRHFARQ128F4252000', 'Musica', 'Sag mir wo wohnen die Engel', 0, 211.61751, 'ARAGY2M1187B9A6712', 'ARAGY2M1187B9A6712', 'eurovision winner'), ('TRHPIRY12903CED4AE', 'O Mein Papa', 'Musical Highlights_Vol.1', 1995, 186.46159, 'ARAGY2M1187B9A6712', 'ARAGY2M1187B9A6712', 'eurovision winner'), ('TRCMWIA12903CED4A9', 'Ponylied', 'Musical Highlights_Vol.1', 1995, 151.17016, 'ARAGY2M1187B9A6712', 'ARAGY2M1187B9A6712', 'eurovision winner'), ('TRCVIEW128F4260948', 'O mein Papa', 'Als Mama noch ein Teeny war', 1995, 181.05424, 'ARAGY2M1187B9A6712', 'ARAGY2M1187B9A6712', 'eurovision winner'), ('TRFRUNJ12903CFF963', 'Was kann schöner sein_ als deine Liebe', 'Wirtschaftswu

We can see that the transaction prevents us from accidentally deleting records we don't want to delete. So we can roll back to the original stage. We can know this rollback work by running the same query after we rollback to see if there is any difference.